test/helpers/value_hook.dart

```diff
@@ -49,48 +223,190 @@ class VHook<T> {
+      if (condition != null) {
+        return condition(val);
+      }
+
+      // No custom condition specified, value recieved -> complete
+      return true;
+    }
+
+    // Throw if already completed with error
+    _assertNotCompletedError();
+
+    // Check current value
+    if (wrappedCond(_value)) return _value;
+
+    // Throw if already completed as no updates are expected
+    _assertNotCompleted();
+
+    // Subscribe to changes and find first that satisfies [condition]
+    Future<T> value = _stream.stream.firstWhere(wrappedCond);
+
+    // Add optional timeout
+    if (timeout != null) value = value.timeout(timeout);
+
+    // Wait for condition to be met
+    return await value;
+  }
+
+  /// Wait for and return the completed value.
+  ///
+  /// When specified [timeout] is waited for completion
+  /// before a [TimeoutException] is thrown.
+  ///
+  /// ```dart
+  /// // Construct VHook with no initial value
+  /// VHook<int> conditionValue = VHook<int>.empty();
   ///
-  /// // This will throw an error if the condition evaluated to false
-  /// conditionValue.expect(equals(true));
+  /// // This function is supposed to call the handler soon
+  /// willCallHandlerAsyncRecurring(() {
+  ///   // These will not cause the [conditionValue] to complete
+  ///   if (someCondition) {
+  ///     conditionValue.set(42);
+  ///   } else {
+  ///     conditionValue.complete(17);
+  ///   }
+  ///
+  ///   // You have so specifically signal completion
+  ///   if (someRarerCondition) {
+  ///     conditionValue.complete();
+  ///   }
+  /// });
+  ///
+  /// // Wait for the conditionValue to be completed
+  /// // If not completed after 5s, throw an error
+  /// await conditionValue.awaitCompletion(timeout: Duration(seconds: 5));
   /// ```
-  Future<bool> awaitValue(
-    Duration? timeout, {
-    Duration pollInterval = Duration.zero,
-    bool raiseOnTimeout = false,
-  }) async {
-    return waitFor(
-      () => _value != null,
-      timeout: timeout,
-      pollInterval: pollInterval,
-      raiseOnTimeout: raiseOnTimeout,
-    );
+  Future<T> awaitCompletion([Duration? timeout]) async {
+    if (timeout != null) {
+      return await _completer.future.timeout(
+        timeout,
+        onTimeout: () => throw TimeoutException('Timed out awaiting value'),
+      );
+    }
+
+    return await _completer.future;
   }
+
+  /// Assert that the value is not yet completed
+  void _assertNotCompleted() {
+    if (_completer.isCompleted) {
+      throw StateError(
+        'VHook is completed with a value that does not satisfy the provided condition',
+      );
+    }
+  }
+
+  /// Assert that value is not yet completed with an error
+  ///
+  /// Will throw [_error] if set
+  void _assertNotCompletedError() {
+    if (_error != null) throw _error!;
+  }
+}
+
+/// Used to encapsulate a value.
+///
+/// This is useful to distinguish  between the value `null`
+/// and 'argument not specified' for optional arguments.
+class VHookValue<T> {
```

Comment:
The name does not really transport much information imo. 
I'd recommend a more general type like `[Value]Container` or `[Value]Wrapper`

Reply:
I wanted to keep the name close to its only use case, but can of course rename this.

Reply:
Maybe just add a `Container` to the name and we are both fine :)

Reply:
`VHookValueContainer` is way to long in my opinion, let's just use `ValueContainer` then